python_modules/dagster/dagster/_core/definitions/auto_materialize_rule_impls.py

```diff
@@ -281,17 +281,22 @@ def passes(
                         AssetKeyPartitionKey(asset_key, record.partition_key)
                     )
 
+        run_ids_with_required_tags = set()
+
         if len(asset_partitions_by_latest_run_id) > 0:
-            run_ids_with_required_tags = (
-                context.legacy_context.instance_queryer.instance.get_run_ids(
-                    filters=RunsFilter(
-                        run_ids=list(asset_partitions_by_latest_run_id.keys()),
-                        tags=self.latest_run_required_tags,
-                    )
+            run_step = int(os.getenv("DAGSTER_ASSET_DAEMON_RUN_TAGS_RUN_FETCH_LIMIT", "1000"))
+
+            required_tag_items = self.latest_run_required_tags.items()
+
+            run_ids_to_fetch = list(asset_partitions_by_latest_run_id.keys())
+            for i in range(0, len(run_ids_to_fetch), run_step):
+                run_ids = run_ids_to_fetch[i : i + run_step]
+                runs = context.legacy_context.instance_queryer.instance.get_runs(
+                    filters=RunsFilter(run_ids=run_ids)
+                )
+                run_ids_with_required_tags.update(
+                    {run.run_id for run in runs if required_tag_items <= run.tags.items()}
```

Comment:
if we just need run ids - can we just query the tags table directly ? is there a good api for that? that should avoid any weird join shenanigans right

Reply:
that is a reasonable point... that is more or less the exact subquery that it is using already

Reply:
er that's not true about the subquery, but still a reasonable point

Reply:
A quick test suggests that for that to work, the logic would need to be something like:
- fetch all rows from run_tags with those run IDs
- do the filtering by tag in memory, return those run IDs

Query planner does not appear to be smart enough to understand the case where there are a small number of run IDs and a large number of candidate runs necessitates fetching by run ID first

Not sure how much that optimization is getting us here other than what's in this PR, other than avoiding needing to fetch the run bodies